src/kernels/webgl/backend_webgl.ts

```diff
@@ -1468,11 +1470,31 @@ export class MathBackendWebGL implements KernelBackend {
   }
 
   addN<T extends Tensor>(tensors: T[]): T {
-    let res = tensors[0];
-    for (let i = 1; i < tensors.length; i++) {
-      res = this.add(res, tensors[i]) as T;
+    if (tensors.length === 1) {
+      return tensors[0];
     }
-    return res;
+
+    // Limit the number of uploaded textures for optimization.
+    if (tensors.length > ENV.get('WEBGL_MAX_TEXTURES_IN_SHADER')) {
+      const midIndex = Math.floor(tensors.length / 2);
+      const leftSide = this.addN(tensors.slice(0, midIndex));
+      const rightSide = this.addN(tensors.slice(midIndex));
+      return this.addN([leftSide, rightSide]);
+    }
+
+    const dtype = tensors
+        .map(t => t.dtype)
+        .reduce((d1, d2) => upcastType(d1, d2));
+    const shapes = tensors.map(t => t.shape);
+    // We can make sure shapes are identical in op level.
+    const usePackedOp = ENV.getBool('WEBGL_PACK_BINARY_OPERATIONS');
```

Comment:
Let's just check `WEBGL_PACK` here rather than checking `WEBGL_PACK_BINARY_OPERATIONS` or creating a new flag.

Reply:
Thanks for pointing out!